### ISSUE TITLE
concurrent map delete and read not panic, read got key of default value

### ISSUE THREAD

**Issue by Cas-pian** [golang/go#17922]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go version go1.7.3 linux/amd64

### What operating system and processor architecture are you using (`go env`)?
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/opt/go_3rdparty"
GORACE=""
GOROOT="/opt/go"
GOTOOLDIR="/opt/go/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build972690274=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"

### What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

concurrent map delete and read should panic, but it didn't panic and then read success with key of default value !
This is the test code:
```
package main

import (
        "fmt"
        "math"
        "time"
)

func main() {
        var m = make(map[int32]bool)

        go func() {
                for {
                        for k, v := range m {
                                if k == 0 {
                                        fmt.Println(k, v)
                                        panic(fmt.Errorf("No! got 0! key should >= 1 !"))
                                }
                                time.Sleep(1 * time.Millisecond)
                        }
                }
        }()
        i := int32(1) // make sure key>=1
        for i < math.MaxInt32 {
                m[i] = true
                delete(m, i)
                i++
                time.Sleep(1 * time.Millisecond)
        }
}
```

### What did you expect to see?
panic info like this:
`fatal error: concurrent map read and map write`

### What did you see instead?
```
0 false
panic: No! got 0! key should >= 1 !

goroutine 5 [running]:
panic(0x490fe0, 0xc42000a3b0)
        /opt/go/src/runtime/panic.go:500 +0x1a1
main.main.func1(0xc420014210)
        /home/caspian/goprojects/src/testmap.go:17 +0x1c8
created by main.main
        /home/caspian/goprojects/src/testmap.go:22 +0x73
exit status 2
```

**Comment by randall77** [comment]
This is fixed in tip, https://go-review.googlesource.com/c/24749/ (treating iteration as read)
Note that this panic is best effort, you should not rely on it.

**Comment by Cas-pian** [comment]
@randall77, Thanks for your reply!   
I get it, this is fixed in the main branch, but 1.7.x release doesn't contain it.

**Comment by davecheney** [comment]
Don't forget this part

> Note that this panic is best effort, you should not rely on it.

If in doubt, run your code under the race detector.

**Comment by Cas-pian** [comment]
@davecheney，OK，thanks very much!

**Comment by bradfitz** [comment]
> I get it, this is fixed in the main branch, but 1.7.x release doesn't contain it.

And it never will. It will be in Go 1.8 and later.
